### DISCUSSION TITLE
How can I get the String from the query ?

### DISCUSSION THREAD

**Discussion by rullyalves** [ExpediaGroup/graphql-kotlin#1373] [Q&A]
Hi, first I would like to thank you for the repository

I have a code similar to this, I'm using spring boot with kotlin and the graphql kotlin spring server library:
  ```kotlin
@Component
class ProfileQuery(private val profileDao: ProfileDao) : Query {
     fun profile(env: DataFetchingEnvironment): List<Profile> {
         return profileDao.getAll()
     }
}

```
I would like to know how I can retrieve the String value from the graphql query, for example:

query {
    profile {
     id
     name
    }
}

**Comment by samuelAndalon** [comment]
Hello, what i would ask is what are you trying to achieve ?

 if you really need to access to the original query string you can use [Instrumentation](https://www.graphql-java.com/documentation/instrumentation)

and the various hooks that it provides:
`beginExecuteOperation`, 
`beginExecutionStrategy`, 
`beginFieldFetch`, 

from those hooks you can access to the `ContextInput` that contains the query

**Comment by rullyalves** [comment]
Hello, so I would like to get the String of the query directly from the DataFetchingEnvironment, is there any way to do this? something like:

```kotlin
@Component
class ProfileQuery(private val profileDao: ProfileDao) : Query {
     fun profile(env: DataFetchingEnvironment): List<Profile> {
         val b = env.getContext.graphqlQuery
         return listOf()
     }
}
```

Can I get this value from ServerRequest, for example?

**Comment by samuelAndalon** [comment]
>Can I get this value from ServerRequest, for example?

in theory at this layer of the execution you should be agnostic of an http request because you are executing a graphQL operation.

unfortunately there is no way to access to the query because at the dataFetcher layer the raw query its not needed. more info here
https://github.com/graphql-java/graphql-java/issues/2373

you could try to use the [AST Document which is available in the DataFetchingEnvironment](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/schema/DataFetchingEnvironment.java#L257)

but again, i would verify what are you trying to achieve, and if is needed rely on the instrumentation hooks. (those actually provide the query string)

**Comment by dariuszkuc** [comment]
You could store raw query in your `GraphQLContext` (factory accepts incoming http request) but similarly to @samuelAndalon I am unsure on the actual use case for doing that.
